samcli/local/apigw/service.py

```diff
@@ -14,6 +14,18 @@
 LOG = logging.getLogger(__name__)
 
 
+class CaseInsensitiveDict(dict):
+
+    def __getitem__(self, key):
+        try:
+            return [v for k, v in self.items() if k.lower() == key.lower()][0]
```

Comment:
I can't really wrap my brain around what is going on here and seems complex for what it should be doing.

Why not have a `__setitem__` where you .lower on insert?

Reply:
I originally did that (have a __setitem__ method that lowercases).  That ended up generating a response with only lowercase headers, which after some thought I decided was a "bigger" change than a minimal change necessary. Since while headers are case-insensitive for comparisons us humans typically look at the case and think that a returned header of "content-type" looks wrong.

So, after some careful consideration realizing that this is only used for local testing, went with find the header as slightly more complicated (it's used only once in the server). While preserving the case of headers.

Reply:
That makes sense. I hadn't considered that. I do agree it would be a strange experience if the Header names magically changed in the request.

Can you refactor this to not use a list comprehension though? I am ok with simple ones but this is fairly complex and hard to parse/understand. We favor readability where we can.

Reply:
Sorry adding one more thing here that I forgot to mention in the comment above. 

Could you also add a comment about why we are only implementing the ‘__getitem__’? This will help future people not be like, what is this madness. You should totally just use ‘__setitem__’ with this as well. :)